vm/src/vector_2.cpp

```diff
@@ -0,0 +1,63 @@
+#include "vector_2.h"
+#include "deserializer.h"
+#include "matrix_3.h"
+#include "serializer.h"
+#include <cassert>
+
+namespace Borges {
+
+Vector_2::Vector_2(std::initializer_list<double> v)
+{
+    assert(v.size() == 2);
+    for (std::size_t i = 0; i < 2; ++i) {
+        v_[i] = v.begin()[i];
+    }
+}
+
+Vector_2::Vector_2(const double v[2])
+{
+    for (std::size_t i = 0; i < 2; ++i) {
+        v_[i] = v[i];
+    }
+}
+
+double
+Vector_2::operator[](std::size_t i) const
+{
+    assert(i < 2);
+    return v_[i];
+}
+
+void
+serialize(Serializer& s, const Vector_2& v)
+{
+    for (std::size_t i = 0; i < 2; ++i) {
+        serialize(s, v[i]);
+    }
+}
+
+void
+deserialize(Deserializer& d, Vector_2* p)
+{
+    double v[2];
+    for (std::size_t i = 0; i < 2; ++i) {
+        v[i] = deserialize<double>(d);
+    }
+    new (p) Vector_2(v);
+}
+
+Vector_2
```

Comment:
why do we need the algebraic relations between the vectors and matrices ? I suggest we stick to the existing Matrix library which works, and convert between Vector_3 and Matrix_4 and Vec3 and Mat4.

Reply:
So we don't accidentally try to use a Vector in a context where a Matrix is expected? For instance, we should only ever pre-multiply a matrix with a vector, so M * v should be allowed, but v * M should not.

In any case, for the purpose of being able to quickly port the existing NURBS tool functionality to the interpreter, I'm fine with keeping support or the existing Matrix library. As we discussed, my proposal is to not further extend the API for these classes for now, and provide conversion functions between these classes and yours.

Reply:
i don't understand your first argument, but it seems moot anyway because of the second argument :)